pyteal/ast/router_test.py

```diff
@@ -0,0 +1,433 @@
+    _a: pt.abi.Uint64,
+    _b: pt.abi.Uint64,
+    _c: pt.abi.Uint64,
+    _d: pt.abi.Uint64,
+    _e: pt.abi.Uint64,
+    _f: pt.abi.Uint64,
+    _g: pt.abi.Uint64,
+    _h: pt.abi.Uint64,
+    _i: pt.abi.Uint64,
+    _j: pt.abi.Uint64,
+    _k: pt.abi.Uint64,
+    _l: pt.abi.Uint64,
+    _m: pt.abi.Uint64,
+    _n: pt.abi.Uint64,
+    _o: pt.abi.Uint64,
+    _p: pt.abi.Uint64,
+    _q: pt.abi.Uint64,
+    _r: pt.abi.Uint64,
+    _s: pt.abi.Uint64,
+    _t: pt.abi.Uint64,
+    *,
+    output: pt.abi.Uint64,
+) -> pt.Expr:
+    return output.set(_t.get())
+
+
+@pt.Subroutine(pt.TealType.none)
+def safe_clear_state_delete():
+    return (
+        pt.If(pt.Txn.sender() == pt.Global.creator_address())
+        .Then(pt.Approve())
+        .Else(pt.Reject())
+    )
+
+
+@pt.ABIReturnSubroutine
+def dummy_doing_nothing():
+    return pt.Seq(pt.Log(pt.Bytes("a message")))
+
+
+@pt.Subroutine(pt.TealType.uint64)
+def returning_u64():
+    return pt.Int(1)
+
+
+@pt.Subroutine(pt.TealType.none)
+def mult_over_u64_and_log(a: pt.Expr, b: pt.Expr):
+    return pt.Log(pt.Itob(a * b))
+
+
+@pt.ABIReturnSubroutine
+def eine_constant(*, output: pt.abi.Uint64):
+    return output.set(1)
+
+
+@pt.ABIReturnSubroutine
+def take_abi_and_log(tb_logged: pt.abi.String):
+    return pt.Log(tb_logged.get())
+
+
+@pt.ABIReturnSubroutine
+def not_registrable(lhs: pt.abi.Uint64, rhs: pt.Expr, *, output: pt.abi.Uint64):
+    return output.set(lhs.get() * rhs)
+
+
+GOOD_SUBROUTINE_CASES: list[pt.ABIReturnSubroutine | pt.SubroutineFnWrapper] = [
+    add,
+    sub,
+    mul,
+    div,
+    mod,
+    qrem,
+    reverse,
+    concat_strings,
+    many_args,
+    safe_clear_state_delete,
+    dummy_doing_nothing,
+    eine_constant,
+    take_abi_and_log,
+]
+
+ON_COMPLETE_CASES: list[pt.EnumInt] = [
+    pt.OnComplete.NoOp,
+    pt.OnComplete.OptIn,
+    pt.OnComplete.ClearState,
+    pt.OnComplete.CloseOut,
+    pt.OnComplete.UpdateApplication,
+    pt.OnComplete.DeleteApplication,
+]
+
+
+def non_empty_power_set(no_dup_list: list):
+    masks = [1 << i for i in range(len(no_dup_list))]
+    for i in range(1, 1 << len(no_dup_list)):
+        yield [elem for mask, elem in zip(masks, no_dup_list) if i & mask]
+
+
+def is_sth_in_oc_list(sth: pt.EnumInt, oc_list: list[pt.EnumInt]):
+    return any(map(lambda x: str(x) == str(sth), oc_list))
+
+
+def assemble_helper(what: pt.Expr) -> pt.TealBlock:
+    assembled, _ = what.__teal__(options)
+    assembled.addIncoming()
+    assembled = pt.TealBlock.NormalizeBlocks(assembled)
+    return assembled
+
+
+def test_parse_conditions():
+    ON_COMPLETE_COMBINED_CASES = non_empty_power_set(ON_COMPLETE_CASES)
+
+    for subroutine, on_completes, is_creation in itertools.product(
+        GOOD_SUBROUTINE_CASES, ON_COMPLETE_COMBINED_CASES, [False, True]
+    ):
+        if not isinstance(subroutine, pt.ABIReturnSubroutine):
+            subroutine = None
+
+        method_sig = subroutine.method_signature() if subroutine else None
```

Comment:
```suggestion
        method_sig = subroutine.method_signature()
```
It looks to me like `subroutine` is never `None`. Am I missing something?

Reply:
oopsies